src/multi_processing/writer.py

```diff
@@ -0,0 +1,248 @@
+        this until termination instruction observed
+    wait_for_jobs()
+        Sleeps until items are observed on the Multiprocessed Queue.
+    handle_jobs(pool_size)
+        Pulls an item from the Multiprocessed Queue if terminate, issues
+        termination, otherwise passes the List containing lists of records
+        to the format method.
+    format_jobs(record_list)
+        Splits jobs into object name & records, then adds to storage & issues
+        instruction to calculate writes.
+    get_job()
+        Retrieve an item from the queue
+    issue_termination()
+        Set the termination flag to True
+    add_to_stored_records(domain_object, records)
+        Add observed data into a dictionary "Domain_Object : All_Records"
+    calculate_writes()
+        Calculate any writes to perform based on currently observed data
+    get_file_num()
+        Increment & return a value recording current file number
+    run_jobs(pool_size)
+        Execute the current job list on the write pool
+    reset_job_list()
+        Clear contents of the job list
+    calculate_residual_writes()
+        Calculate any remaining writes after terminate has been observed
+    """
+
+    def __init__(self, write_queue, max_file_size, file_builder):
+        """ Assign write queue, maximum file size, and the file builder to
+        use. Additionally, set starting/deafult values for all records,
+
+        Parameters
+        ----------
+        write_queue : Multiprocessing Queue
+            Shared, multiprocessing safe, queue holding lists of lists of
+            records
+        max_file_size : int
+            Value representing the maximum number of records per file
+        file_builder : File_Builder
+            Instantiated file builder, pre-configured to output the necessary
+            file extension.
+        """
+
+        self.write_queue = write_queue
+
+        self.all_records = []
+        self.file_number = 0
+        self.max_size = int(max_file_size)
+
+        self.job_list = []
+        self.terminate = False
+        self.file_builder = file_builder
+
+    def start(self, pool_size):
+        """ Begin the cycle of waiting for, handling, and running jobs,
+        continuing this until an instruction to terminate is observed. Once
+        observed, calculate any residual write to be made and terminate.
+
+        Parameters
+        ----------
+        pool_size : int
+            The number of processes running in the generator's pool
+        """
+
+        while not self.terminate:
+            self.wait_for_jobs()
+            self.handle_jobs(pool_size)
+            self.run_jobs(pool_size)
+            self.reset_job_list()
+
+        self.calculate_residual_writes()
+        self.run_jobs(pool_size)
+
+    def wait_for_jobs(self):
+        """ Sleep until records are on the queue """
+
+        while self.write_queue.empty():
+            time.sleep(1)
```

Comment:
what happens if this never receives a generated record to write? For example if there is a generator or validation issue? Do we shut this down gracefully?

Reply:
No, we drag it down kicking and screaming with task manager/pkill at the minute. Issue raised: #58